File: java/src/test/java/org/demo/java/EmailAddressTests.java
package org.demo.java;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class EmailAddressTests {

    @Test
    public void parsing() {
        assertEquals(
                new EmailAddress("fred", "example.com"),
                EmailAddress.parse("fred@example.com")
        );
    }

    @Test
    public void parsingFailures() {
        assertThrows(
                IllegalArgumentException.class,
                () -> EmailAddress.parse("@"));
    }

}

File: kotlin/src/test/kotlin/org/demo/kotlin/EmailAddressTests.kt
package org.demo.kotlin

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class EmailAddressTests {
    @Test
    fun parsing() {
        Assertions.assertEquals(
            EmailAddress("fred", "example.com"),
            EmailAddress.parse("fred@example.com")
        )
    }

    @Test
    fun parsingFailures() {
        Assertions.assertThrows(
            IllegalArgumentException::class.java
        ) { EmailAddress.parse("@") }
    }
}